### PR TITLE
Issue #191 in CogniCrypt: Provider Detection analysis

### LINKED ISSUE #126: Add Provider Detection analysis

### PR DESCRIPTION
**Pull request by enriozuni**
Closed #126 
Closed  enriozuni/https://github.com/eclipse-cognicrypt/CogniCrypt/#191

This pull request closes issue 191 in CogniCrypt, which implements the _Provider Detection_ analysis. 
The description of the issue can be read by clicking [here](https://github.com/eclipse-cognicrypt/CogniCrypt/issues/191).

### CODE REVIEW

**Comment by johspaeth** [comment]
Thanks, the code looks ok to me now. One last thing: Please move the functionality into a separate class.

Before merging, I also would like to have tested the functionality with real BC-provider rules.

**Comment by enriozuni** [comment]
> Thanks, the code looks ok to me now. One last thing: Please move the functionality into a separate class.
> 
> Before merging, I also would like to have tested the functionality with real BC-provider rules.

@johspaeth I moved the functionality into a new class. Currently, I am working with the test cases. I will push the changes as soon as I am finished.

File: CryptoAnalysis/src/main/java/crypto/ProviderDetection.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+		Val forwardQueryVal = forwardQuery.var();
+		Value value = forwardQueryVal.value();
+		Type valueType = value.getType();
+		String valueTypeString = valueType.toString();
+		
+		if(valueTypeString.matches("org.bouncycastle.jce.provider.BouncyCastleProvider")) {
+			provider = "BC";
+			
+			//choosing the rules depending on Provider detected
+			File rule = new File(".\\.\\.\\test\\resources\\"+provider+"\\"+refName+".cryptslbin");
```
Comment (kruegers) [inline CryptoAnalysis/src/main/java/crypto/ProviderDetection.java]: What does ".\\.\\.\\" do?

File: CryptoAnalysis/src/main/java/crypto/ProviderDetection.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+		Map<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> map = backwardQueryResults.getAllocationSites();
+		Entry<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> entry = map.entrySet().iterator().next();
+		ForwardQuery forwardQuery = entry.getKey();
+		
+		Val forwardQueryVal = forwardQuery.var();
+		Value value = forwardQueryVal.value();
+		Type valueType = value.getType();
+		String valueTypeString = valueType.toString();
+		
+		if(valueTypeString.matches("org.bouncycastle.jce.provider.BouncyCastleProvider")) {
```
Comment (kruegers) [inline CryptoAnalysis/src/main/java/crypto/ProviderDetection.java]: The following code seems rather generic and applicable regardless of which provider is selected. Why do you keep going only when the provider is BC?

File: CryptoAnalysis/src/main/java/crypto/ProviderDetection.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		solver.debugOutput();
+		
+		Map<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> map = backwardQueryResults.getAllocationSites();
+		Entry<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> entry = map.entrySet().iterator().next();
+		ForwardQuery forwardQuery = entry.getKey();
+		
+		Val forwardQueryVal = forwardQuery.var();
+		Value value = forwardQueryVal.value();
+		Type valueType = value.getType();
+		String valueTypeString = valueType.toString();
```
Comment (kruegers) [inline CryptoAnalysis/src/main/java/crypto/ProviderDetection.java]: This method and chooseProviderWhenProviderTypeString are virtually the same, starting here. Rather call it here and return its return value. Makes maintenance easier.

File: CryptoAnalysis/src/main/java/crypto/ProviderDetection.java (hunk 1 above)
Comment (enriozuni) [reply CryptoAnalysis/src/main/java/crypto/ProviderDetection.java]: Each dot (.) navigates to its parent directory. This way, it can navigate to the _Provider_ directory (if it exists)

File: CryptoAnalysis/src/main/java/crypto/ProviderDetection.java (hunk 2 above)
Comment (enriozuni) [reply CryptoAnalysis/src/main/java/crypto/ProviderDetection.java]: I know it is not the most elegant solution. I wanted to extract the provider name using Soot and Boomerang, (for example BC) and then use its name to navigate to the directory and select the corresponding rules, but I was not able to find a way. The closest I could get was to extract the Provider type and assign a value to the _provider_ param, so I can then use it to navigate to the directory where the rules are.

File: CryptoAnalysis/src/main/java/crypto/ProviderDetection.java (hunk 3 above)
Comment (enriozuni) [reply CryptoAnalysis/src/main/java/crypto/ProviderDetection.java]: That is true. I will make the corresponding changes. Thanks

File: CryptoAnalysis/src/main/java/crypto/ProviderDetection.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+				return null;
+			}
+		};
+		
+		//Submit query to the solver.
+		BackwardBoomerangResults<NoWeight> backwardQueryResults = solver.solve(query);
+		solver.debugOutput();
+		
+		Map<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> map = backwardQueryResults.getAllocationSites();
+		Entry<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> entry = map.entrySet().iterator().next();
```
Comment (johspaeth) [inline CryptoAnalysis/src/main/java/crypto/ProviderDetection.java]: That's a complicated way of iterating over a map.

You can use 
`for(entry : map.entrySet()){}`

Comment (johspaeth) [reply CryptoAnalysis/src/main/java/crypto/ProviderDetection.java]: Also you just are just using the first entry. What about the others?

Comment (enriozuni) [reply CryptoAnalysis/src/main/java/crypto/ProviderDetection.java]: I tested this scenario using Boomerang and the size of the _map_ is always 1 since there is only one allocation site for the _Provider_ variable. I also just tested your suggestion about the iteration and the output is the same since there is always only 1 element in the _map_.  In my case, I was just getting the first entry from the map, and extracting the corresponding _key_ value. Anyway, I can use your iteration if you want. Thanks

Comment (johspaeth) [reply CryptoAnalysis/src/main/java/crypto/ProviderDetection.java]: A code such as 
`Cipher.getInstance(...,(someChoice() ? "BouncyCastle" : "Default"))`
would yield a size of the map larger than one. And your code *randomly* picks one of the providers as the order of the elements in the map depend on the hashcode of the objects, which may differ in each run of the JVM. We have to prevent such non-deterministic choices.

Please think of a solution and better throw an exception in the case there are multiple providers. Then we at least know where the code crashes. Thanks.

Comment (enriozuni) [reply CryptoAnalysis/src/main/java/crypto/ProviderDetection.java]: @johspaeth I had not thought about this case. Thanks for exposing this scenario. I will make the required changes to the code.

File: CryptoAnalysis/src/main/java/crypto/providerdetection/ProviderDetection.java [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+	
+	private String getProviderType(Value vl) {
+		Type type = vl.getType();
+//		System.out.println("Soot type: "+type);
+		String strType = type.toString();
+		return strType;
+	}
+	
+	
+	private String getProviderWhenTypeString(Value vl, Body body) {
```
Comment (enriozuni) [inline CryptoAnalysis/src/main/java/crypto/providerdetection/ProviderDetection.java]: @kruegers This method does provider detection when the provider is of type `java.lang.String`.

File: CryptoAnalysis/.project [hunk 6]
```diff
@@ -10,6 +10,11 @@
 			<arguments>
 			</arguments>
 		</buildCommand>
+		<buildCommand>
```
Comment (kruegers) [inline CryptoAnalysis/.project]: Please revert the changes in this file.

Comment (enriozuni) [reply CryptoAnalysis/.project]: I did it. Thanks for noticing that. Maybe that is why my previous Shippable build was failing
